include/swift/Reflection/ReflectionContext.h

```diff
@@ -789,61 +789,150 @@ class ReflectionContext
+      return false;
+    auto typeResult = readTypeFromMetadata(result.getValue());
+    if (!typeResult)
+      return false;
+    *OutInstanceTR = typeResult;
+    *OutInstanceAddress = RemoteAddress(*pointerval);
+    return true;
+  }
+
+  bool getDynamicTypeAndAddressErrorExistential(
+      RemoteAddress object, const TypeRef **OutInstanceTR,
+      RemoteAddress *OutInstanceAddress, bool *IsBridgedError = nullptr) {
+    auto result = readMetadataAndValueErrorExistential(object);
+    if (!result)
+      return false;
+
+    auto typeResult =
+        readTypeFromMetadata(result->MetadataAddress.getAddressData());
+    if (!typeResult)
+      return false;
+
+    *OutInstanceTR = typeResult;
+    *OutInstanceAddress = result->PayloadAddress;
+    if (IsBridgedError)
+      *IsBridgedError = result->IsBridgedError;
+
+    return true;
+  }
+
+  bool
+  getDynamicTypeAndAddressOpaqueExistential(RemoteAddress object,
+                                            const TypeRef **OutInstanceTR,
+                                            RemoteAddress *OutInstanceAddress) {
+    auto result = readMetadataAndValueOpaqueExistential(object);
+    if (!result)
+      return false;
+
+    auto typeResult =
+        readTypeFromMetadata(result->MetadataAddress.getAddressData());
+    if (!typeResult)
+      return false;
+
+    *OutInstanceTR = typeResult;
+    *OutInstanceAddress = result->PayloadAddress;
+
+    return true;
+  }
+
   bool projectExistential(RemoteAddress ExistentialAddress,
                           const TypeRef *ExistentialTR,
                           const TypeRef **OutInstanceTR,
                           RemoteAddress *OutInstanceAddress,
                           remote::TypeInfoProvider *ExternalTypeInfo) {
-    if (ExistentialTR == nullptr)
+    auto ExistentialRecordTI =
+        getRecordTypeInfo(ExistentialTR, ExternalTypeInfo);
+    if (ExistentialRecordTI == nullptr)
       return false;
 
-    auto ExistentialTI = getTypeInfo(ExistentialTR, ExternalTypeInfo);
-    if (ExistentialTI == nullptr)
+    switch (ExistentialRecordTI->getRecordKind()) {
+    case RecordKind::ClassExistential:
+      return getDynamicTypeAndAddressClassExistential(
+          ExistentialAddress, OutInstanceTR, OutInstanceAddress);
+
+    case RecordKind::OpaqueExistential: {
+      return getDynamicTypeAndAddressOpaqueExistential(
+          ExistentialAddress, OutInstanceTR, OutInstanceAddress);
+    }
+    case RecordKind::ErrorExistential: {
+      return getDynamicTypeAndAddressErrorExistential(
+          ExistentialAddress, OutInstanceTR, OutInstanceAddress);
+    }
+    default:
       return false;
+    }
+  }
+
+  bool projectExistentialForLLDB(RemoteAddress ExistentialAddress,
+                                 const TypeRef *ExistentialTR,
+                                 const TypeRef **OutInstanceTR,
+                                 RemoteAddress *OutInstanceAddress) {
+    auto IsClass = [](const TypeRef *TypeResult) {
```

Comment:
Since we expect TypeResult to be non-null, we might as well pass in a `const TypeRef &` here?